From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: David Slovikosky <davidslovikosky@gmail.com>
Date: Tue, 9 Jun 2020 00:10:03 -0700
Subject: [PATCH] Fix missing chunks due to integer overflow

This patch fixes a bug in the WorldChunkManagerTheEnd class where the distance
from 0,0 squared overflows the maximum size of an integer. The overflow leads
to hard chunk borders around 370,000 blocks from 0,0. After this cutoff there
is a few hundred thousand block gap before end land resuming to generate at
530,000 blocks from spawn. This is due to the integer flipping back and forth.

The fix for the issue is quite simple, casting chunk coordinates to longs
allows the distance calculation to avoid overflow and work as intended.

diff --git a/src/main/java/net/minecraft/world/level/biome/TheEndBiomeSource.java b/src/main/java/net/minecraft/world/level/biome/TheEndBiomeSource.java
index af006c2e45e0a14367a0bc850c319024c6b82024..063369d3a64b4afc9cc6e1d20360900595e1a05f 100644
--- a/src/main/java/net/minecraft/world/level/biome/TheEndBiomeSource.java
+++ b/src/main/java/net/minecraft/world/level/biome/TheEndBiomeSource.java
@@ -75,7 +75,9 @@ public class TheEndBiomeSource extends BiomeSource {
         int l = j / 2;
         int i1 = i % 2;
         int j1 = j % 2;
-        float f = 100.0F - Mth.sqrt((float) (i * i + j * j)) * 8.0F;
+        // Paper start - cast ints to long to avoid integer overflow
+        float f = 100.0F - Mth.sqrt((long) i * (long) i + (long) j * (long) j) * 8.0F;
+        // Paper end
 
         f = Mth.clamp(f, -100.0F, 80.0F);
 
